File: python/src/threads/team/imported.py
"""An import stores a log's bytes without its appends' hooks, so in the same transaction the
index is folded again from the logs: each imported branch's wake rows, and every team the log
belongs to, rebuilt whole. A team whose lead is not stored yet has no rows until its lead is
imported (that import rebuilds it). A log that breaks rule 43 with the stored team is refused."""

import time

from pydantic.experimental.missing_sentinel import MISSING

from threads.log import ParseError, TeamOpenedEvent, ThreadStartedEvent
from threads.result import Err
from threads.store import receipts, sql, wakes
from threads.store.conn import Conn
from threads.store.indexing import known
from threads.store.started import opened_threads
from threads.store.verify import VerifiedLog
from threads.team.rebuild import refold_team


class _UndoError(Exception):
    def __init__(self, error: ParseError) -> None:
        self.error = error


def import_indexed(
    conn: Conn, log: VerifiedLog, tenant: str, dropped_ref: str | None
) -> ParseError | None:
    """Stores the log's segments and folds the index rows they change, in one transaction."""
    try:
        with sql.transaction(conn):
            error = sql.insert_segments(conn, log, tenant, dropped_ref) or _index(conn, log, tenant)
            if error is not None:
                raise _UndoError(error)
    except _UndoError as undo:
        return undo.error
    return None


def _index(conn: Conn, log: VerifiedLog, tenant: str) -> ParseError | None:
    now = int(time.time() * 1000)
    for s in log.segments:
        events = known([e for e, _ in s.events])
        wakes.refold(conn, s.header.branch_id, events)
        receipts.rebuild_ui(conn, tenant, events, now)
    for team in _teams_of(conn, log, tenant):
        rebuilt = refold_team(conn, tenant, team)
        if isinstance(rebuilt, Err) and rebuilt.error.code != "not_found":
            return rebuilt.error
    return None


def _teams_of(conn: Conn, log: VerifiedLog, tenant: str) -> list[str]:
    """The teams the log's thread belongs to, from its opening event: the team it leads, the
    team whose log it is, and, for a member, its lead's team."""
    events = log.segments[0].events
    first = events[0][0] if events else None
    if isinstance(first, TeamOpenedEvent):
        return [first.data.team]
    if not isinstance(first, ThreadStartedEvent):
        return []
    own = [] if first.data.team is MISSING else [first.data.team.id]
    parent = first.data.parent
    if parent is MISSING or parent.relation != "team_member":
        return own
    lead = next((o for o in opened_threads(conn, tenant) if o.thread_id == parent.thread_id), None)
    if lead is None or not isinstance(lead.event, ThreadStartedEvent):
        return own
    team = lead.event.data.team
    return own if team is MISSING else [*own, team.id]

File: typescript/packages/core/src/team/imported.ts
import type { TeamId } from "../log";
import { ok, type Result } from "../result";
import type { LogStore } from "../store";
import { knownOf } from "../store/indexing";
import { openedThreads } from "../store/started";
import { refoldWakes } from "../store/wakes";
import type { VerifiedLog } from "../verify";
import type { LogError } from "../verify/error";
import { rebuildTeamIndex } from "./rebuild";

// An import stores a log's bytes without its appends' hooks, so in the same transaction the
// index is folded again from the logs: each imported branch's wake rows, and every team the log
// belongs to, rebuilt whole. A team whose lead is not stored yet has no rows until its lead is
// imported (that import rebuilds it). A log that breaks rule 43 with the stored team is refused.

/** Folds the index rows an imported `log` changes; runs inside the import's transaction. */
export function indexImported(
  store: LogStore,
  log: VerifiedLog,
): Result<void, LogError> {
  for (const s of log.segments)
    refoldWakes(store.driver, s.header.branch_id, knownOf(s.events));
  const teams = teamsOf(store, log);
  if (!teams.ok) return teams;
  for (const team of teams.value) {
    const rebuilt = rebuildTeamIndex(store, team);
    if (!rebuilt.ok && rebuilt.error.code !== "not_found") return rebuilt;
  }
  return ok(undefined);
}

/**
 * The teams the log's thread belongs to, from its opening event: the team it leads, the team
 * whose log it is, and, for a member, its lead's team.
 */
function teamsOf(
  store: LogStore,
  log: VerifiedLog,
): Result<readonly TeamId[], LogError> {
  const first = log.segments[0]?.events[0];
  const e = first?.kind === "event" ? first.event : undefined;
  if (e?.type === "team_opened") return ok([e.data.team]);
  if (e?.type !== "thread_started") return ok([]);
  const own = e.data.team === undefined ? [] : [e.data.team.id];
  const parent = e.data.parent;
  if (parent?.relation !== "team_member") return ok(own);
  const opened = openedThreads(store.driver, store.tenant);
  if (!opened.ok) return opened;
  const lead = opened.value.find((o) => o.threadId === parent.thread_id);
  const team =
    lead?.event.type === "thread_started"
      ? lead.event.data.team?.id
      : undefined;
  return ok(team === undefined ? own : [...own, team]);
}
